src/olympia/versions/tasks.py

```diff
@@ -112,11 +113,95 @@ def delete_preview_files(pk, **kw):
 
 @task
 @use_primary_db
-def extract_version_to_git(version_id, author_id=None, note=None):
-    """Extract a `File` into our git storage backend."""
+def extract_addon_to_git(addon_pk):
+    addon = Addon.unfiltered.get(pk=addon_pk)
+    log.info(
+        'Starting extraction of addon "{}" to git storage.'.format(addon_pk)
+    )
+
+    if addon.git_extraction_is_in_progress:
+        log.info('Aborting extraction of addon "{}" to git storage because it '
+                 'is already in progress.'.format(addon_pk))
+        return
+
+    # "Lock" this add-on so that the git extraction can only run once and tasks
+    # for extracting related versions are delayed until this task has finished.
+    git_storage, created = GitExtraction.objects.update_or_create(
+        addon=addon,
+        defaults={'in_progress': True},
+    )
+
+    # Filter out versions that are already present in the git storage.
+    versions = addon.versions.filter(git_hash='').order_by('created')
+
+    for version in versions:
+        try:
+            log.info('Starting extraction of version "{}" for addon "{}" to '
+                     'git storage.'.format(version.pk, addon_pk))
+
+            extract_version_to_git(
+                version.pk,
+                # We do not want to trigger an infinite loop so an error will
+                # be thrown if there is a broken git reference.
+                stop_on_broken_ref=True,
+                # The `extract_version_to_git` task also checks the lock we set
+                # above so we need to tell it to skip this check here.
+                can_be_delayed=False,
+            )
+
+            log.info('Ending extraction of version "{}" for addon "{}" to '
```

Comment:
Can't work if using tasks, obviously. This, and the `git_storage.update(in_progress=False)`, likely need to be in a `chord` callback... Actually, a `chain` even: since we're not obeying the lock, we probably want the extractions to happen in sequence, not in parallel.

Reply:
So we would build a gigantic chain of tasks for each version? That will probably speed up the whole process but it makes things more complicated since we'll have to handle errors in the chain (for instance, to reset the lock).

Reply:
Yeah, a big chain is not ideal for sure, I'm open to alternatives :) We definitely can't do everything in a single task...

Reply:
> We definitely can't do everything in a single task...

That, I tend to agree but that's how we did it to migrate _via_ the command, right?

In any case, I think the big chain isn't ideal but sounds a bit better than doing everything in a single task. I can't think of any other way...

Reply:
I can think of three alternatives... the first two I wouldn't recommend, but here for the sake of the discussion:
- Trigger a single task that processes a single or more versions and then once it's done, triggers a new task to do the next bit.
- Trigger `x` tasks but don't bother with a callback, instead check on status elsewhere in a loop.

My favorite approach:
- Trigger `x` tasks of `y` versions (essentially what you're doing now, except you're processing _all_ versions instead of a fixed `y` number). This is similar to the approach taken by reindex and code search, which both use a `chord` with a callback to signal that it's finished. The trick is to find a reasonable `y` value.